javaslang/src/main/java/javaslang/collection/HashMultimap.java

```diff
@@ -149,24 +160,92 @@ private Builder(ContainerType containerType, SerializableSupplier<Traversable<?>
         return (HashMultimap<K, V>) map;
     }
 
-    private HashMultimap(Map<K, Traversable<V>> back, ContainerType containerType, SerializableSupplier<Traversable<?>> emptyContainer) {
-        super(back, containerType, emptyContainer);
+    // -- non-static API
+
+    @Override
+    public boolean containsKey(K key) {
+        return back.containsKey(key);
     }
 
     @Override
-    protected <K2, V2> Map<K2, V2> emptyMapSupplier() {
-        return HashMap.empty();
+    public ContainerType getContainerType() {
+        return containerType;
+    }
+
+    @Override
+    public Option<Traversable<V>> get(K key) {
+        return back.get(key);
+    }
+
+    @Override
+    public Set<K> keySet() {
+        return back.keySet();
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    protected <K2, V2> HashMultimap<K2, V2> emptyInstance() {
-        return new HashMultimap<>(HashMap.empty(), getContainerType(), emptyContainer);
+    public <K2, V2> HashMultimap<K2, V2> flatMap(BiFunction<? super K, ? super V, ? extends Iterable<Tuple2<K2, V2>>> mapper) {
+        Objects.requireNonNull(mapper, "mapper is null");
+        return foldLeft(emptyInstance(), (acc, entry) -> {
+            for (Tuple2<? extends K2, ? extends V2> mappedEntry : mapper.apply(entry._1, entry._2)) {
+                acc = acc.put(mappedEntry);
+            }
+            return acc;
+        });
     }
 
     @Override
-    protected <K2, V2> HashMultimap<K2, V2> createFromMap(Map<K2, Traversable<V2>> back) {
-        return new HashMultimap<>(back, getContainerType(), emptyContainer);
+    public int size() {
+        return back.foldLeft(0, (s, t) -> s + t._2.size());
```

Comment:
I think it will not be quadratic because valid containers do store their size. It will be linear over keys.

However, storing the size might be a viable solution. We should do that in a separate PR when this re-design is finished. We currently have too many changes in one PR.